dot/parachain/prospective-parachains/messages.go

```diff
@@ -0,0 +1,292 @@
+		return 2, ppm.inner, nil
+
+	case GetHypotheticalMembership:
+		return 3, ppm.inner, nil
+
+	case GetMinimumRelayParents:
+		return 4, ppm.inner, nil
+	}
+
+	return 0, nil, scale.ErrUnsupportedVaryingDataTypeValue
+}
+
+func (ppm ProspectiveParachainsMessage) Value() (value any, err error) {
+	_, value, err = ppm.IndexValue()
+	return
+}
+
+func (ppm ProspectiveParachainsMessage) ValueAt(index uint) (value any, err error) {
+	switch index {
+	case 0:
+		return *new(IntroduceSecondedCandidate), nil
+
+	case 1:
+		return *new(CandidateBacked), nil
+
+	case 2:
+		return *new(GetBackableCandidates), nil
+
+	case 3:
+		return *new(GetHypotheticalMembership), nil
+
+	case 4:
+		return *new(GetMinimumRelayParents), nil
+
+	}
+	return nil, scale.ErrUnknownVaryingDataTypeValue
+}
+
+// NewCollatorProtocolMessage returns a new collator protocol message varying data type
+func NewProspectiveParachainMessage() ProspectiveParachainsMessage {
+	return ProspectiveParachainsMessage{}
+}
+
+type IntroduceSecondedCandidate struct {
+	IntroduceSecondedCandidateRequest
+	Sender chan bool
+}
+
+// IntroduceSecondedCandidateRequest Request introduction of a seconded candidate into the prospective parachains
+// subsystem.
+type IntroduceSecondedCandidateRequest struct {
+	// The para-id of the candidate.
+	CandidatePara parachaintypes.ParaID
+	// The candidate receipt itself.
+	CandidateReceipt parachaintypes.CommittedCandidateReceipt
+	// The persisted validation data of the candidate.
+	PersistedValidationData parachaintypes.PersistedValidationData
+}
+
+// CandidateBacked Inform the Prospective Parachains Subsystem that a previously introduced candidate
+// has been backed. This requires that the candidate was successfully introduced in
+// the past.
+type CandidateBacked struct {
+	ParaId        parachaintypes.ParaID
+	CandidateHash parachaintypes.CandidateHash
+}
+
+// GetBackableCandidates Try getting N backable candidate hashes along with their relay parents for the given
+// parachain, under the given relay-parent hash, which is a descendant of the given ancestors.
+// Timed out ancestors should not be included in the collection.
+// N should represent the number of scheduled cores of this ParaId.
+// A timed out ancestor frees the cores of all of its descendants, so if there's a hole in the
+// supplied ancestor path, we'll get candidates that backfill those timed out slots first. It
+// may also return less/no candidates, if there aren't enough backable candidates recorded.
+type GetBackableCandidates struct {
+	Hash      common.Hash
+	ParaId    parachaintypes.ParaID
+	N         uint32
+	Ancestors Ancestors
+	Sender    chan []struct {
+		CandidateHash parachaintypes.CandidateHash
+		RelayParent   common.Hash
+	}
+}
+
+// Ancestors A collection of ancestor candidates of a parachain.
+// TODO: is this correct?  Rust code has 'Ancestors' as pub type Ancestors = HashSet<CandidateHash>;
+type Ancestors []parachaintypes.CandidateHash
```

Comment:
yep it is correct!

Reply:
I don't think Rust type `HashSet<CandidateHash>` is same as `[]parachaintypes.CandidateHash` in Go.
We implement set using map in Golang.

you can change it to `type Ancestors map[parachaintypes.CandidateHash]struct{}`